java/ql/src/semmle/code/java/security/SpelInjection.qll

```diff
@@ -0,0 +1,145 @@
+/** Provides classes to reason about SpEL injection attacks. */
+
+import java
+import semmle.code.java.dataflow.DataFlow
+import semmle.code.java.dataflow.ExternalFlow
+import semmle.code.java.dataflow.FlowSources
+
+/** A data flow sink for unvalidated user input that is used to construct SpEL expressions. */
+abstract class SpelExpressionEvaluationSink extends DataFlow::ExprNode { }
+
+private class SpelExpressionEvaluationModel extends SinkModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "org.springframework.expression;Expression;true;getValue;;;Argument[-1];spel",
+        "org.springframework.expression;Expression;true;getValueTypeDescriptor;;;Argument[-1];spel",
+        "org.springframework.expression;Expression;true;getValueType;;;Argument[-1];spel",
+        "org.springframework.expression;Expression;true;setValue;;;Argument[-1];spel"
+      ]
+  }
+}
+
+/** Default sink for SpEL injection vulnerabilities. */
+private class DefaultSpelExpressionEvaluationSink extends SpelExpressionEvaluationSink {
+  DefaultSpelExpressionEvaluationSink() {
+    exists(MethodAccess ma |
+      sinkNode(this, "spel") and
+      this.asExpr() = ma.getQualifier() and
+      not exists(SafeEvaluationContextFlowConfig config |
+        config.hasFlowTo(DataFlow::exprNode(ma.getArgument(0)))
```

Comment:
Not sure if this is within the scope of this pull request, but should this rather be modeled as sanitizer to reduce false negatives?

Reply:
Could you provide an example of the kind of false negatives you would expect to prevent with this change?

Reply:
I don't have a specific example but I was thinking of something like the following. Though I have not tested it, and it might not compile:
```java
public void unsafe(Socket socket) throws IOException {
    InputStream in = socket.getInputStream();

    byte[] bytes = new byte[1024];
    int n = in.read(bytes);
    String input = new String(bytes, 0, n);

    StandardEvaluationContext context = new StandardEvaluationContext();
    evaluate(input, context);
}

public void safe(String s) {
    SimpleEvaluationContext context = SimpleEvaluationContext.forReadWriteDataBinding().build();
    evaluate(s, context);
}

private void evaluate(String s, EvaluationContext context) {
    Expression expression = PARSER.parseExpression(s);
    // Won't be detected since there is also flow from `safe(String)` to here
    expression.getValue(context);
}
```

But maybe such false negatives are not actually easily avoidable.

Reply:
Thank you for the clarification. Given the amount of effort needed to cover such a case (since the current data-flow implementation doesn't allow to do this well), and that we're not aware of a real use-case for which this change would add value, we're going to leave it as it is for the time being. Happy to revisit this if false negatives following this pattern are found in real codebases, though.